File: ffi/bindings/dotnet/rodbus-tests/IntegrationTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

using rodbus;
using System;
using System.Collections.Generic;
using System.Linq;

namespace rodbus_tests
{
    class WriteHandler : IWriteHandler
    {
        public WriteResult WriteSingleCoil(ushort index, bool value, Database database)
        {
            if (database.UpdateCoil(index, value))
            {
                return WriteResult.SuccessInit();
            }
            else
            {
                return WriteResult.ExceptionInit(rodbus.ModbusException.IllegalDataAddress);
            }
        }

        public WriteResult WriteSingleRegister(ushort index, ushort value, Database database)
        {
            if (database.UpdateHoldingRegister(index, value))
            {
                return WriteResult.SuccessInit();
            }
            else
            {
                return WriteResult.ExceptionInit(rodbus.ModbusException.IllegalDataAddress);
            }
        }

        public WriteResult WriteMultipleCoils(ushort start, ICollection<BitValue> it, Database database)
        {
            foreach (var bit in it)
            {
                if (!database.UpdateCoil(bit.Index, bit.Value))
                {
                    return WriteResult.ExceptionInit(rodbus.ModbusException.IllegalDataAddress);
                }
            }

            return WriteResult.SuccessInit();
        }

        public WriteResult WriteMultipleRegisters(ushort start, ICollection<RegisterValue> it, Database database)
        {
            foreach (var bit in it)
            {
                if (!database.UpdateHoldingRegister(bit.Index, bit.Value))
                {
                    return WriteResult.ExceptionInit(rodbus.ModbusException.IllegalDataAddress);
                }
            }

            return WriteResult.SuccessInit();
        }
    }

    [TestClass]
    public class IntegrationTest
    {
        private static readonly byte UNIT_ID = 1;
        private static readonly ushort NUM_POINTS = 10;
        // we use 50001 here since it's a large enough port it doesn't require root on Linux
        private static readonly string ENDPOINT = "127.0.0.1";
        private static readonly ushort PORT = 50001;
        private static readonly RequestParam param = new RequestParam(UNIT_ID, TimeSpan.FromSeconds(1));

        static void TestReadDiscreteInputs(ClientChannel client)
        {
            var result = client.ReadDiscreteInputs(param, new AddressRange(2, 3)).Result.ToList();
            Assert.AreEqual(3, result.Count);
            Assert.AreEqual(2, result[0].Index);
            Assert.AreEqual(false, result[0].Value);
            Assert.AreEqual(3, result[1].Index);
            Assert.AreEqual(true, result[1].Value);
            Assert.AreEqual(4, result[2].Index);
            Assert.AreEqual(false, result[2].Value);

            try
            {
                client.ReadDiscreteInputs(param, new AddressRange(9, 2)).Wait();
                Assert.Fail("reading invalid address range did not fail");
            }
            catch (AggregateException ex)
            {
                Assert.AreEqual(RequestError.ModbusExceptionIllegalDataAddress, (ex.InnerException as RequestException).error);
            }
        }

        static void TestReadInputRegisters(ClientChannel client)
        {
            var result = client.ReadInputRegisters(param, new AddressRange(3, 3)).Result.ToList();
            Assert.AreEqual(3, result.Count);
            Assert.AreEqual(3, result[0].Index);
            Assert.AreEqual(0, result[0].Value);
            Assert.AreEqual(4, result[1].Index);
            Assert.AreEqual(42, result[1].Value);
            Assert.AreEqual(5, result[2].Index);
            Assert.AreEqual(0, result[2].Value);

            try
            {
                client.ReadInputRegisters(param, new AddressRange(10, 1)).Wait();
                Assert.Fail("reading invalid address range did not fail");
            }
            catch(AggregateException ex)
            {
                Assert.AreEqual(RequestError.ModbusExceptionIllegalDataAddress, (ex.InnerException as RequestException).error);
            }
        }

        static void TestWriteSingleCoil(ClientChannel client)
        {
            client.WriteSingleCoil(param, new BitValue(1, true)).Wait();

            var result = client.ReadCoils(param, new AddressRange(0, 2)).Result.ToList();
            Assert.AreEqual(2, result.Count);
            Assert.AreEqual(0, result[0].Index);
            Assert.AreEqual(false, result[0].Value);
            Assert.AreEqual(1, result[1].Index);
            Assert.AreEqual(true, result[1].Value);
        }

        static void TestWriteSingleRegister(ClientChannel client)
        {
            client.WriteSingleRegister(param, new RegisterValue(1, 22)).Wait();

            var result = client.ReadHoldingRegisters(param, new AddressRange(0, 2)).Result.ToList();
            Assert.AreEqual(2, result.Count);
            Assert.AreEqual(0, result[0].Index);
            Assert.AreEqual(0, result[0].Value);
            Assert.AreEqual(1, result[1].Index);
            Assert.AreEqual(22, result[1].Value);
        }

        static void TestWriteMultipeCoils(ClientChannel client)
        {
            client.WriteMultipleCoils(param, 0, new List<bool> { true, false, true }).Wait();

            var result = client.ReadCoils(param, new AddressRange(0, 3)).Result.ToList();
            Assert.AreEqual(3, result.Count);
            Assert.AreEqual(0, result[0].Index);
            Assert.AreEqual(true, result[0].Value);
            Assert.AreEqual(1, result[1].Index);
            Assert.AreEqual(false, result[1].Value);
            Assert.AreEqual(2, result[2].Index);
            Assert.AreEqual(true, result[2].Value);
        }

        static void TestWriteMultipeRegisters(ClientChannel client)
        {
            client.WriteMultipleRegisters(param, 0, new List<ushort> { 0xCAFE, 21, 0xFFFF }).Wait();

            var result = client.ReadHoldingRegisters(param, new AddressRange(0, 3)).Result.ToList();
            Assert.AreEqual(3, result.Count);
            Assert.AreEqual(0, result[0].Index);
            Assert.AreEqual(0xCAFE, result[0].Value);
            Assert.AreEqual(1, result[1].Index);
            Assert.AreEqual(21, result[1].Value);
            Assert.AreEqual(2, result[2].Index);
            Assert.AreEqual(0xFFFF, result[2].Value);
        }

        [TestMethod]
        public void ClientAndServerCommunication()
        {
            var runtime = new Runtime(new RuntimeConfig());
            var map = new DeviceMap();
            map.AddEndpoint(UNIT_ID, new WriteHandler(), db =>
            {
                for (ushort i = 0; i < NUM_POINTS; ++i)
                {
                    db.AddCoil(i, false);
                    db.AddDiscreteInput(i, false);
                    db.AddHoldingRegister(i, 0);
                    db.AddInputRegister(i, 0);
                }
            });

            var server = Server.CreateTcp(runtime, ENDPOINT, 100, map, DecodeLevel.Nothing());
            var client = ClientChannel.CreateTcp(runtime, ENDPOINT, PORT, 10, new RetryStrategy(), DecodeLevel.Nothing());

            // set a unique pattern to test reads
            server.UpdateDatabase(UNIT_ID, db =>
            {
                db.UpdateDiscreteInput(3, true);
                db.UpdateInputRegister(4, 42);
            });

            TestReadDiscreteInputs(client);
            TestReadInputRegisters(client);
            TestWriteSingleCoil(client);
            TestWriteSingleRegister(client);
            TestWriteMultipeCoils(client);
            TestWriteMultipeRegisters(client);
        }
    }
}

File: ffi/bindings/java/rodbus-tests/src/test/java/io/stepfunc/rodbus/tests/IntegrationTest.java
package io.stepfunc.rodbus.tests;

import io.stepfunc.rodbus.*;
import io.stepfunc.rodbus.ModbusException;
import io.stepfunc.rodbus.Runtime;
import org.joou.*;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.parallel.Execution;

import java.time.Duration;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.joou.Unsigned.*;

class IntegrationTest {
    static final UByte UNIT_ID = ubyte(1);
    static final Duration TIMEOUT = Duration.ofSeconds(1);
    static final int NUM_POINTS = 10;
    // we use 50001 here since it's a large enough port it doesn't require root on Linux
    static final String ENDPOINT = "127.0.0.1:50001";

    static class TestWriteHandler implements WriteHandler {
        @Override
        public WriteResult writeSingleCoil(UShort index, boolean value, Database database) {
            if (database.updateCoil(index, value)) {
                return WriteResult.successInit();
            } else {
                return WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
            }
        }

        @Override
        public WriteResult writeSingleRegister(UShort index, UShort value, Database database) {
            if (database.updateHoldingRegister(index, value)) {
                return WriteResult.successInit();
            } else {
                return WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
            }
        }

        @Override
        public WriteResult writeMultipleCoils(UShort start, List<BitValue> it, Database database) {
            for (BitValue bit : it) {
                if (!database.updateCoil(bit.index, bit.value)) {
                    return WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
                }
            }

            return WriteResult.successInit();
        }

        @Override
        public WriteResult writeMultipleRegisters(UShort start, List<RegisterValue> it, Database database) {
            for (RegisterValue register : it) {
                if (!database.updateHoldingRegister(register.index, register.value)) {
                    return WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
                }
            }

            return WriteResult.successInit();
        }
    }

    @Test
    void clientAndServerCommunication() throws ExecutionException, InterruptedException {
        final RuntimeConfig runtimeConfig = new RuntimeConfig();
        Runtime runtime = new Runtime(runtimeConfig);

        final DeviceMap deviceMap = new DeviceMap();
        deviceMap.addEndpoint(UNIT_ID, new TestWriteHandler(), db -> {
            for(int i = 0; i < NUM_POINTS; i++) {
                db.addCoil(ushort(i), false);
                db.addDiscreteInput(ushort(i), false);
                db.addHoldingRegister(ushort(i), ushort(0));
                db.addInputRegister(ushort(i), ushort(0));
            }
        });

        final Server server = Server.createTcp(runtime, ENDPOINT, ushort(100), deviceMap, DecodeLevel.nothing());
        final ClientChannel client = ClientChannel.createTcp(runtime, ENDPOINT, ushort(10), new RetryStrategy(), DecodeLevel.nothing());

        // Set a unique pattern to test reads
        server.updateDatabase(UNIT_ID, db -> {
            db.updateDiscreteInput(ushort(3), true);
            db.updateInputRegister(ushort(4), ushort(42));
        });

        testReadDiscreteInputs(client);
        testReadInputRegisters(client);
        testWriteSingleCoil(client);
        testWriteSingleRegister(client);
        testWriteMultipleCoils(client);
        testWriteMultipleRegisters(client);
    }

    private void testReadDiscreteInputs(ClientChannel client) throws ExecutionException, InterruptedException {
        RequestParam param = new RequestParam(UNIT_ID, TIMEOUT);
        AddressRange range = new AddressRange(ushort(2), ushort(3));

        List<BitValue> result = client.readDiscreteInputs(param, range).toCompletableFuture().get();

        assertThat(result).hasSize(3);
        assertThat(result.get(0).index).isEqualTo(ushort(2));
        assertThat(result.get(0).value).isFalse();
        assertThat(result.get(1).index).isEqualTo(ushort(3));
        assertThat(result.get(1).value).isTrue();
        assertThat(result.get(2).index).isEqualTo(ushort(4));
        assertThat(result.get(2).value).isFalse();

        // ======

        assertThatThrownBy(() -> {
            range.start = ushort(9);
            range.count = ushort(2);
            client.readDiscreteInputs(param, range).toCompletableFuture().get();
        }).isInstanceOf(ExecutionException.class).extracting("getCause.error").isEqualTo(RequestError.MODBUS_EXCEPTION_ILLEGAL_DATA_ADDRESS);
    }

    private void testReadInputRegisters(ClientChannel client) throws ExecutionException, InterruptedException {
        RequestParam param = new RequestParam(UNIT_ID, TIMEOUT);
        AddressRange range = new AddressRange(ushort(3), ushort(3));

        List<RegisterValue> result = client.readInputRegisters(param, range).toCompletableFuture().get();

        assertThat(result).hasSize(3);
        assertThat(result.get(0).index).isEqualTo(ushort(3));
        assertThat(result.get(0).value).isEqualTo(ushort(0));
        assertThat(result.get(1).index).isEqualTo(ushort(4));
        assertThat(result.get(1).value).isEqualTo(ushort(42));
        assertThat(result.get(2).index).isEqualTo(ushort(5));
        assertThat(result.get(2).value).isEqualTo(ushort(0));

        // ======

        assertThatThrownBy(() -> {
            range.start = ushort(10);
            range.count = ushort(1);
            client.readDiscreteInputs(param, range).toCompletableFuture().get();
        }).isInstanceOf(ExecutionException.class).extracting("getCause.error").isEqualTo(RequestError.MODBUS_EXCEPTION_ILLEGAL_DATA_ADDRESS);
    }

    private void testWriteSingleCoil(ClientChannel client) throws ExecutionException, InterruptedException {
        RequestParam param = new RequestParam(UNIT_ID, TIMEOUT);
        BitValue bit = new BitValue(ushort(1), true);

        client.writeSingleCoil(param, bit).toCompletableFuture().get();

        // ======

        AddressRange range = new AddressRange(ushort(0), ushort(2));

        List<BitValue> readResult = client.readCoils(param, range).toCompletableFuture().get();

        assertThat(readResult).hasSize(2);
        assertThat(readResult.get(0).index).isEqualTo(ushort(0));
        assertThat(readResult.get(0).value).isFalse();
        assertThat(readResult.get(1).index).isEqualTo(ushort(1));
        assertThat(readResult.get(1).value).isTrue();
    }

    private void testWriteSingleRegister(ClientChannel client) throws ExecutionException, InterruptedException {
        RequestParam param = new RequestParam(UNIT_ID, TIMEOUT);
        RegisterValue register = new RegisterValue(ushort(1), ushort(22));

        client.writeSingleRegister(param, register).toCompletableFuture().get();

        // ======

        AddressRange range = new AddressRange(ushort(0), ushort(2));

        List<RegisterValue> readResult = client.readHoldingRegisters(param, range).toCompletableFuture().get();

        assertThat(readResult).hasSize(2);
        assertThat(readResult.get(0).index).isEqualTo(ushort(0));
        assertThat(readResult.get(0).value).isEqualTo(ushort(0));
        assertThat(readResult.get(1).index).isEqualTo(ushort(1));
        assertThat(readResult.get(1).value).isEqualTo(ushort(22));
    }

    private void testWriteMultipleCoils(ClientChannel client) throws ExecutionException, InterruptedException {
        RequestParam param = new RequestParam(UNIT_ID, TIMEOUT);

        client.writeMultipleCoils(param, ushort(0), Arrays.asList(true, false, true)).toCompletableFuture().get();

        // ======

        AddressRange range = new AddressRange(ushort(0), ushort(3));

        List<BitValue> readResult = client.readCoils(param, range).toCompletableFuture().get();

        assertThat(readResult).hasSize(3);
        assertThat(readResult.get(0).index).isEqualTo(ushort(0));
        assertThat(readResult.get(0).value).isTrue();
        assertThat(readResult.get(1).index).isEqualTo(ushort(1));
        assertThat(readResult.get(1).value).isFalse();
        assertThat(readResult.get(2).index).isEqualTo(ushort(2));
        assertThat(readResult.get(2).value).isTrue();
    }

    private void testWriteMultipleRegisters(ClientChannel client) throws ExecutionException, InterruptedException {
        RequestParam param = new RequestParam(UNIT_ID, TIMEOUT);

        client.writeMultipleRegisters(param, ushort(0), Arrays.asList(ushort(0xCAFE), ushort(21), ushort(0xFFFF))).toCompletableFuture().get();

        // ======

        AddressRange range = new AddressRange(ushort(0), ushort(3));

        List<RegisterValue> readResult = client.readHoldingRegisters(param, range).toCompletableFuture().get();

        assertThat(readResult).hasSize(3);
        assertThat(readResult.get(0).index).isEqualTo(ushort(0));
        assertThat(readResult.get(0).value).isEqualTo(ushort(0xCAFE));
        assertThat(readResult.get(1).index).isEqualTo(ushort(1));
        assertThat(readResult.get(1).value).isEqualTo(ushort(21));
        assertThat(readResult.get(2).index).isEqualTo(ushort(2));
        assertThat(readResult.get(2).value).isEqualTo(ushort(0xFFFF));
    }
}
